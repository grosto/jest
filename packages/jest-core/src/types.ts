/**
 * Copyright (c) Facebook, Inc. and its affiliates. All Rights Reserved.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import {Context} from 'jest-runtime';
import {Test} from 'jest-runner';
import {Config, TestResult} from '@jest/types';

export type TestRunData = Array<{
  context: Context;
  matches: {
    allTests: number;
    tests: Array<Test>;
    total?: number;
    stats?: Stats;
  };
}>;

type TestPathCaseStats = Record<keyof (TestPathCases), number>;

export type TestPathCaseWithPathPatternStats = Record<
  keyof (TestPathCasesWithPathPattern),
  number
>;

export type Stats = TestPathCaseStats | TestPathCaseWithPathPatternStats;

export type TestPathCases = {
  roots: (path: Config.Path) => boolean;
  testMatch: (path: Config.Path) => boolean;
  testPathIgnorePatterns: (path: Config.Path) => boolean;
  testRegex: (path: Config.Path) => boolean;
};

export type TestPathCasesWithPathPattern = TestPathCases & {
  testPathPattern: (path: Config.Path) => boolean;
};

// TODO: Obtain this from @jest/reporters once its been migrated
export type ReporterOnStartOptions = {
  estimatedTime: number;
  showStatus: boolean;
};

export type Reporter = {
  onTestResult: (
    test: Test,
    testResult: TestResult.TestResult,
    aggregatedResult: TestResult.AggregatedResult,
  ) => Promise<void>;
  onRunStart: (
    results: TestResult.AggregatedResult,
    options: ReporterOnStartOptions,
  ) => Promise<void>;
  onTestStart: (test: Test) => Promise<void>;
  onRunComplete: (
    contexts: Set<Context>,
    results: TestResult.AggregatedResult,
  ) => Promise<void>;
  getLastError: () => Error;
};
