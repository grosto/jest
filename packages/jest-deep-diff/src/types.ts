/**
 * Copyright (c) Facebook, Inc. and its affiliates. All Rights Reserved.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import {NewPlugin} from 'pretty-format/src/types';

// options
export type DeepDiffOptions = Omit<FormatOptions, 'plugins'> & {
  plugins?: Array<Plugin>;
};

export type FormatOptions = Partial<FormatOptionsNormalized>;

export type DiffOptionsColor = (arg: string) => string; // subset of Chalk type

export type FormatOptionsNormalized = {
  aAnnotation: string;
  aColor: DiffOptionsColor;
  aIndicator: string;
  bAnnotation: string;
  bColor: DiffOptionsColor;
  bIndicator: string;
  changeColor: DiffOptionsColor;
  changeLineTrailingSpaceColor: DiffOptionsColor;
  commonColor: DiffOptionsColor;
  commonIndicator: string;
  commonLineTrailingSpaceColor: DiffOptionsColor;
  contextLines: number;
  emptyFirstOrLastLinePlaceholder: string;
  expand: boolean;
  includeChangeCounts: boolean;
  omitAnnotationLines: boolean;
  patchColor: DiffOptionsColor;
  serialize: (val: unknown) => string;
  plugins: Array<FormatPlugin>;
};

// plugins
export interface Plugin {
  serialize: Serialize;
  test: PluginTest;
  diff: CustomDiff;
  format: CustomFormat;
}

export type PluginTest = (a: unknown) => boolean;

export interface FormatPlugin {
  test: PluginTest;
  format: CustomFormat;
}

export interface DiffPlugin {
  test: PluginTest;
  diff: CustomDiff;
}

type Serialize = NewPlugin['serialize'];
export interface SerializePlugin {
  test: PluginTest;
  serialize: Serialize;
}

// DiffObject
export type Path = unknown;

export enum Kind {
  EQUAL,
  UPDATED,
  INSERTED,
  DELETED,
  UNEQUAL_TYPE,
}
export interface DiffObject<T1 = unknown, T2 = T1> {
  kind: Kind;
  path?: Path;
  a: T1;
  b: T2;
  childDiffs?: Array<DiffObject>;
}

export type Memos = {
  a: Map<unknown, number>;
  b: Map<unknown, number>;
  position: number;
};

export type DiffFunc<T1 = unknown, T2 = T1> = (
  a: T1,
  b: T2,
  path: Path | undefined,
  memos?: Memos,
) => DiffObject<T1, T2>;

export type CustomDiff = (
  a: unknown,
  b: unknown,
  path: Path | undefined,
  memos: Memos | undefined,
  diff: DiffFunc,
) => DiffObject<unknown, unknown>;

// format

export type LineGenerationOptions = {
  plugins: Array<FormatPlugin>;
};

export type Format<T1 = unknown, T2 = T1> = (
  diff: DiffObject<T1, T2>,
  context: Context,
  options: LineGenerationOptions,
) => Array<Line>;

export type FormatComplexDiff<T1 = unknown, T2 = T1> = (
  diff: DiffObject<T1, T2>,
  context: Context,
  options: LineGenerationOptions,
  format: Format,
) => Array<Line>;

export type CustomFormat = FormatComplexDiff<unknown, unknown>;

export enum LineType {
  COMMON,
  INSERTED,
  DELETED,
}

export interface Context {
  indent?: string;
  prefix?: string;
  sufix?: string;
  skipSerialize?: boolean;
}

export interface Line {
  type: LineType;
  prefix: string;
  val: unknown;
  suffix: string;
  indent: string;
  skipSerialize?: boolean;
}
