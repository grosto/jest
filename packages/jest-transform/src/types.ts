/**
 * Copyright (c) Facebook, Inc. and its affiliates. All Rights Reserved.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import {RawSourceMap} from 'source-map';
import {Config} from '@jest/types';

export type ShouldInstrumentOptions = Pick<
  Config.GlobalConfig,
  | 'collectCoverage'
  | 'collectCoverageFrom'
  | 'collectCoverageOnlyFrom'
  | 'coverageProvider'
> & {
  changedFiles?: Set<Config.Path>;
};

export type Options = ShouldInstrumentOptions &
  Partial<{
    isCoreModule: boolean;
    isInternalModule: boolean;
  }>;

// This is fixed in source-map@0.7.x, but we can't upgrade yet since it's async
interface FixedRawSourceMap extends Omit<RawSourceMap, 'version'> {
  version: number;
}

export type TransformedSource = {
  code: string;
  map?: FixedRawSourceMap | string | null;
};

export type TransformResult = {
  code: string;
  originalCode: string;
  mapCoverage: boolean;
  sourceMapPath: string | null;
};

export type TransformOptions = {
  instrument: boolean;
};

export type CacheKeyOptions = {
  config: Config.ProjectConfig;
  instrument: boolean;
  rootDir: string;
};

export interface Transformer {
  canInstrument?: boolean;
  createTransformer?: (options?: any) => Transformer;

  getCacheKey?: (
    fileData: string,
    filePath: Config.Path,
    configStr: string,
    options: CacheKeyOptions,
  ) => string;

  process: (
    sourceText: string,
    sourcePath: Config.Path,
    config: Config.ProjectConfig,
    options?: TransformOptions,
  ) => string | TransformedSource;
}
