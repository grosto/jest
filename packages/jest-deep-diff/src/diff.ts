import getType = require('jest-get-type');
import {DiffObject, DiffPlugin, Memos, Path} from './types';
import {wrapIfHasCircularity} from './complex/circularObjects';
import {createEqual, createUnequalType, createUpdated} from './diffObject';
import {diffObjects} from './complex/object';
import {diffArrays} from './complex/array';
import {diffMaps} from './complex/map';
import {
  diffBooleans,
  diffDates,
  diffErrors,
  diffFunctions,
  diffNumbers,
  diffRegExps,
  diffStrings,
} from './primitives';

// eslint-disable-next-line @typescript-eslint/ban-types
type PrimitiveWrapper = String | Number | Boolean;

function diff(
  a: unknown,
  b: unknown,
  path?: Path,
  memos?: Memos,
  customDiffs: Array<DiffPlugin> = [],
): DiffObject {
  if (Object.is(a, b)) return createEqual(a, b, path);

  const diffWithCustomDiffs = (
    a: unknown,
    b: unknown,
    path?: Path,
    memos?: Memos,
  ) => diff(a, b, path, memos, customDiffs);

  for (const customDiff of customDiffs) {
    if (customDiff.test(a) || customDiff.test(b)) {
      return customDiff.diff(a, b, path, memos, diffWithCustomDiffs);
    }
  }

  const aType = getType(a);

  const bType = getType(b);
  if (aType !== bType) {
    return createUnequalType(a, b, path);
  }

  if (a instanceof Error) {
    return diffErrors(a, b as Error, path);
  }

  switch (aType) {
    case 'string':
      return diffStrings(a as string, b as string, path);
    case 'number':
      return diffNumbers(a as number, b as number, path);
    case 'boolean':
      return diffBooleans(a as boolean, b as boolean, path);
    case 'date':
      return diffDates(a as Date, b as Date, path);
    case 'regexp':
      return diffRegExps(a as RegExp, b as RegExp, path);
    case 'function':
      return diffFunctions(a as Function, b as Function, path);
  }

  // adapted from node assert
  // Use memos to handle cycles.
  if (memos === undefined) {
    memos = {
      a: new Map(),
      b: new Map(),
      position: 0,
    };
  } else {
    // We prevent up to two map.has(x) calls by directly retrieving the value
    // and checking for undefined. The map can only contain numbers, so it is
    // safe to check for undefined only.
    const bMemoA = memos.a.get(a);
    const bMemoB = memos.b.get(b);
    if (bMemoA !== undefined || bMemoB !== undefined) {
      return bMemoA === bMemoB
        ? createEqual(
            wrapIfHasCircularity(a, bMemoA),
            wrapIfHasCircularity(b, bMemoB),
            path,
          )
        : createUpdated(
            wrapIfHasCircularity(a, bMemoA),
            wrapIfHasCircularity(b, bMemoB),
            path,
          );
    }
    memos.position++;
  }

  memos.a.set(a, memos.position);
  memos.b.set(b, memos.position);

  if (!Object.is(Object.getPrototypeOf(a), Object.getPrototypeOf(b))) {
    return createUnequalType(a, b, path);
  }

  // TODO: DOM Node

  if (a instanceof Number || a instanceof String || a instanceof Boolean) {
    return Object.is(a.valueOf(), (b as PrimitiveWrapper).valueOf())
      ? createEqual(a, b, path)
      : createUpdated(a, b, path);
  }

  const diffWithMemosAndCustomDiffs = (a: unknown, b: unknown, path: Path) =>
    diffWithCustomDiffs(a, b, path, memos);

  switch (aType) {
    case 'array':
      return diffArrays(
        a as Array<unknown>,
        b as Array<unknown>,
        path,
        diffWithMemosAndCustomDiffs,
      );
    case 'object':
      return diffObjects(
        a as Record<string | number | symbol, unknown>,
        b as Record<string | number | symbol, unknown>,
        path,
        diffWithMemosAndCustomDiffs,
      );
    case 'map':
      return diffMaps(
        a as Map<unknown, unknown>,
        b as Map<unknown, unknown>,
        path,
        diffWithMemosAndCustomDiffs,
      );
    // case 'set':
    //   return compareSets(a as Set<unknown>, b as Set<unknown>, path, deepDiff);
    default:
      throw new Error('oopsie ' + aType);
  }
}

export default diff;
