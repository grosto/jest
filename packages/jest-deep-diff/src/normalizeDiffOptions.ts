/**
 * Copyright (c) Facebook, Inc. and its affiliates. All Rights Reserved.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

// adapted from jest-diff
import chalk = require('chalk');
import prettyFormat = require('pretty-format');

import {FormatOptions, FormatOptionsNormalized} from './types';
import {wrappedCircularSerializer} from './complex/circularObjects';

export const noColor = (string: string): string => string;

const DIFF_CONTEXT_DEFAULT = 5;

const OPTIONS_DEFAULT: FormatOptionsNormalized = {
  aAnnotation: 'Expected',
  aColor: chalk.green,
  aIndicator: '-',
  bAnnotation: 'Received',
  bColor: chalk.red,
  bIndicator: '+',
  changeColor: chalk.inverse,
  changeLineTrailingSpaceColor: noColor,
  commonColor: chalk.dim,
  commonIndicator: ' ',
  commonLineTrailingSpaceColor: noColor,
  contextLines: DIFF_CONTEXT_DEFAULT,
  emptyFirstOrLastLinePlaceholder: '',
  expand: true,
  includeChangeCounts: false,
  omitAnnotationLines: false,
  patchColor: chalk.yellow,
  plugins: [],
  serialize: (val: unknown) =>
    prettyFormat(val, {
      plugins: [wrappedCircularSerializer],
    }),
};

const getContextLines = (contextLines?: number): number =>
  typeof contextLines === 'number' &&
  Number.isSafeInteger(contextLines) &&
  contextLines >= 0
    ? contextLines
    : DIFF_CONTEXT_DEFAULT;

// Pure function returns options with all properties.
export const normalizeDiffOptions = (
  options: FormatOptions = {},
): FormatOptionsNormalized => ({
  ...OPTIONS_DEFAULT,
  ...options,
  contextLines: getContextLines(options.contextLines),
});
