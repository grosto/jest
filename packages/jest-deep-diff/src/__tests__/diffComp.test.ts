/**
 * Copyright (c) Facebook, Inc. and its affiliates. All Rights Reserved.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

// adapted from jest-diff/__tests__/diff.test.ts
import chalk = require('chalk');
import oldDiff from 'jest-diff';
import diff from '../';
import {noColor} from '../normalizeDiffOptions';
import {DeepDiffOptions} from '../types';

const optionsCounts: DeepDiffOptions = {
  includeChangeCounts: false,
};

// Use in toBe assertions for comparison lines.
const optionsBe: DeepDiffOptions = {
  aColor: noColor,
  bColor: noColor,
  commonColor: noColor,
  omitAnnotationLines: true,
};
const unexpandedBe: DeepDiffOptions = {
  ...optionsBe,
  expand: false,
};
const expandedBe: DeepDiffOptions = {
  ...optionsBe,
  expand: true,
};

// Use for toMatchSnapshot assertions.
const unexpanded = {expand: false};
const expanded = {expand: true};

const elementSymbol = Symbol.for('react.element');

describe('different types', () => {
  [
    [1, 'a', 'number', 'string'],
    [{}, 'a', 'object', 'string'],
    [[], 2, 'array', 'number'],
    [null, undefined, 'null', 'undefined'],
    [() => {}, 3, 'function', 'number'],
  ].forEach(values => {
    const a = values[0];
    const b = values[1];

    test(`'${String(a)}' and '${String(b)}'`, () => {
      expect(diff(a, b, expandedBe)).toBe(oldDiff(a, b, expandedBe));
    });
  });
});

describe('no visual difference', () => {
  [
    ['a', 'a'],
    [{}, {}],
    [[], []],
    [
      [1, 2],
      [1, 2],
    ],
    [11, 11],
    [NaN, NaN],
    [Number.NaN, NaN],
    /* TBD:These two functions are not considered equal by expect.toEqual but
    jest-diff shows that there are no visual differences. IMHO,
    behaviour should be to show two idential outputs but still
    mark them as updated */
    // [() => {}, () => {}],
    [null, null],
    [undefined, undefined],
    [false, false],
    [{a: 1}, {a: 1}],
    [{a: {b: 5}}, {a: {b: 5}}],
  ].forEach(values => {
    test(`'${JSON.stringify(values[0])}' and '${JSON.stringify(
      values[1],
    )}'`, () => {
      const a = values[0];
      const b = values[1];
      expect(diff(a, b, expandedBe)).toBe(oldDiff(a, b, expandedBe));
    });
  });

  test.skip('Map key order should be irrelevant', () => {
    const arg1 = new Map([
      [1, 'foo'],
      [2, 'bar'],
    ]);
    const arg2 = new Map([
      [2, 'bar'],
      [1, 'foo'],
    ]);

    expect(diff(arg1, arg2, expandedBe)).toBe(oldDiff(arg1, arg2, expandedBe));
  });

  test.skip('Set value order should be irrelevant', () => {
    const arg1 = new Set([1, 2]);
    const arg2 = new Set([2, 1]);

    expect(diff(arg1, arg2, expandedBe)).toBe(oldDiff(arg1, arg2, expandedBe));
  });
});

describe('oneline strings', () => {
  [
    ['ab', 'aa'],
    ['123456789', '234567890'],
    ['oneline', 'multi\nline'],
    ['multi\nline', 'oneline'],
  ].forEach(([a, b]) => {
    test(a + ' ' + b, () => {
      expect(diff(a, b, optionsCounts)).toBe(oldDiff(a, b, optionsCounts));
    });
  });
});

// Some of the following assertions seem complex, but compare to alternatives:
// * toBe instead of toMatchSnapshot:
//   * to avoid visual complexity of escaped quotes in expected string
//   * to omit Expected/Received heading which is an irrelevant detail
// * join lines of expected string instead of multiline string:
//   * to avoid ambiguity about indentation in diff lines

describe('multiline strings', () => {
  const a = `line 1
line 2
line 3
line 4`;
  const b = `line 1
line  2
line 3
line 4`;
  test('(unexpanded)', () => {
    expect(diff(a, b, unexpandedBe)).toBe(oldDiff(a, b, unexpandedBe));
  });
  test('(expanded)', () => {
    expect(diff(a, b, expandedBe)).toBe(oldDiff(a, b, expandedBe));
  });
});

describe('objects', () => {
  const a = {a: {b: {c: 5}}};
  const b = {a: {b: {c: 6}}};

  test('(unexpanded)', () => {
    expect(diff(a, b, unexpandedBe)).toBe(oldDiff(a, b, unexpandedBe));
  });
  test('(expanded)', () => {
    expect(diff(a, b, expandedBe)).toBe(oldDiff(a, b, expandedBe));
  });
});

test('numbers', () => {
  expect(diff(1, 2, optionsBe)).toBe(oldDiff(1, 2, optionsBe));
});

test('-0 and 0', () => {
  expect(diff(-0, 0, optionsBe)).toBe(oldDiff(-0, 0, optionsBe));
});

test('booleans', () => {
  expect(diff(false, true, optionsBe)).toBe(oldDiff(false, true, optionsBe));
});

describe('multiline string non-snapshot', () => {
  // For example, CLI output
  // toBe or toBe for a string isnâ€™t enclosed in double quotes.
  const a = `Options:
--help, -h  Show help                            [boolean]
--bail, -b  Exit the test suite immediately upon the first
            failing test.                        [boolean]`;
  const b = `Options:
  --help, -h  Show help                            [boolean]
  --bail, -b  Exit the test suite immediately upon the first
              failing test.                        [boolean]`;

  test('(unexpanded)', () => {
    expect(diff(a, b, unexpandedBe)).toBe(oldDiff(a, b, unexpandedBe));
  });
  test('(expanded)', () => {
    expect(diff(a, b, expandedBe)).toBe(oldDiff(a, b, expandedBe));
  });
});

describe('multiline string snapshot', () => {
  // For example, CLI output
  // A snapshot of a string is enclosed in double quotes.
  const a = `"
Options:
--help, -h  Show help                            [boolean]
--bail, -b  Exit the test suite immediately upon the first
            failing test.                        [boolean]"`;
  const b = `"
Options:
  --help, -h  Show help                            [boolean]
  --bail, -b  Exit the test suite immediately upon the first
              failing test.                        [boolean]"`;

  test('(unexpanded)', () => {
    expect(diff(a, b, unexpandedBe)).toEqual(oldDiff(a, b, unexpandedBe));
  });
  test('(expanded)', () => {
    expect(diff(a, b, expandedBe)).toEqual(oldDiff(a, b, expandedBe));
  });
});

describe('React elements', () => {
  const a = {
    $$typeof: elementSymbol,
    props: {
      children: 'Hello',
      className: 'fun',
    },
    type: 'div',
  };
  const b = {
    $$typeof: elementSymbol,
    props: {
      children: 'Goodbye',
      className: 'fun',
    },
    type: 'div',
  };

  test('(unexpanded)', () => {
    expect(diff(a, b, unexpandedBe)).toBe(oldDiff(a, b, unexpandedBe));
  });
  test('(expanded)', () => {
    expect(diff(a, b, expandedBe)).toBe(oldDiff(a, b, expandedBe));
  });
});

describe.skip('multiline string as value of object property', () => {
  describe('(non-snapshot)', () => {
    const a = {
      id: 'J',
      points: '0.5,0.460\n0.25,0.875',
    };
    const b = {
      id: 'J',
      points: '0.5,0.460\n0.5,0.875\n0.25,0.875',
    };

    test('(unexpanded)', () => {
      expect(diff(a, b, unexpandedBe)).toBe(oldDiff(a, b, unexpandedBe));
    });
    test('(expanded)', () => {
      expect(diff(a, b, expandedBe)).toBe(oldDiff(a, b, unexpandedBe));
    });
  });
});

describe.skip('indentation in JavaScript structures', () => {
  const searching = '';
  const object = {
    descending: false,
    fieldKey: 'what',
  };
  const a = {
    searching,
    sorting: object,
  };
  const b = {
    searching,
    sorting: [object],
  };

  describe('from less to more', () => {
    test('(unexpanded)', () => {
      expect(diff(a, b, unexpandedBe)).toBe(oldDiff(a, b, unexpandedBe));
    });
    test('(expanded)', () => {
      expect(diff(a, b, expandedBe)).toBe(oldDiff(a, b, expandedBe));
    });
  });

  describe('from more to less', () => {
    test('(unexpanded)', () => {
      expect(diff(a, b, unexpandedBe)).toBe(oldDiff(a, b, unexpandedBe));
    });
    test('(expanded)', () => {
      expect(diff(a, b, expandedBe)).toBe(oldDiff(a, b, expandedBe));
    });
  });
});

describe.skip('color of text', () => {
  const searching = '';
  const object = {
    descending: false,
    fieldKey: 'what',
  };
  const a = {
    searching,
    sorting: object,
  };
  const b = {
    searching,
    sorting: [object],
  };
  const received = diff(a, b, expanded);

  test('(expanded)', () => {
    expect(received).toBe(oldDiff(a, b, expanded));
  });
  test('(unexpanded)', () => {
    // Expect same result, unless diff is long enough to require patch marks.
    expect(diff(a, b, unexpanded)).toBe(received);
  });
});

describe.skip('indentation in React elements (non-snapshot)', () => {
  const leaf = {
    $$typeof: elementSymbol,
    props: {
      children: ['text'],
    },
    type: 'span',
  };
  const a = {
    $$typeof: elementSymbol,
    props: {
      children: [leaf],
    },
    type: 'span',
  };
  const b = {
    $$typeof: elementSymbol,
    props: {
      children: [
        {
          $$typeof: elementSymbol,
          props: {
            children: [leaf],
          },
          type: 'strong',
        },
      ],
    },
    type: 'span',
  };

  describe('from less to more', () => {
    test('(unexpanded)', () => {
      expect(diff(a, b, unexpandedBe)).toBe(oldDiff(a, b, unexpandedBe));
    });
    test('(expanded)', () => {
      expect(diff(a, b, expandedBe)).toBe(oldDiff(a, b, expandedBe));
    });
  });

  describe('from more to less', () => {
    test('(unexpanded)', () => {
      expect(diff(a, b, unexpandedBe)).toBe(oldDiff(a, b, unexpandedBe));
    });
    test('(expanded)', () => {
      expect(diff(a, b, expandedBe)).toBe(oldDiff(a, b, expandedBe));
    });
  });
});

describe.skip('outer React element (non-snapshot)', () => {
  const a = {
    $$typeof: elementSymbol,
    props: {
      children: 'Jest',
    },
    type: 'h1',
  };
  const b = {
    $$typeof: elementSymbol,
    props: {
      children: [
        a,
        {
          $$typeof: elementSymbol,
          props: {
            children: 'Delightful JavaScript Testing',
          },
          type: 'h2',
        },
      ],
    },
    type: 'header',
  };

  describe('from less to more', () => {
    test('(unexpanded)', () => {
      expect(diff(a, b, unexpandedBe)).toBe(oldDiff(a, b, unexpandedBe));
    });
    test('(expanded)', () => {
      expect(diff(a, b, expandedBe)).toBe(oldDiff(a, b, expandedBe));
    });
  });

  describe('from more to less', () => {
    test('(unexpanded)', () => {
      expect(diff(b, a, unexpandedBe)).toBe(oldDiff(a, b, unexpandedBe));
    });
    test('(expanded)', () => {
      expect(diff(b, a, expandedBe)).toBe(oldDiff(a, b, expandedBe));
    });
  });
});

describe.skip('trailing newline in multiline string not enclosed in quotes', () => {
  const a = ['line 1', 'line 2', 'line 3'].join('\n');
  const b = a + '\n';

  describe('from less to more', () => {
    test('(unexpanded)', () => {
      expect(diff(a, b, unexpandedBe)).toBe(oldDiff(a, b, unexpandedBe));
    });
    test('(expanded)', () => {
      expect(diff(a, b, expandedBe)).toBe(oldDiff(a, b, expandedBe));
    });
  });

  describe('from more to less', () => {
    test('(unexpanded)', () => {
      expect(diff(b, a, unexpandedBe)).toBe(oldDiff(a, b, unexpanded));
    });
    test('(expanded)', () => {
      expect(diff(b, a, expandedBe)).toBe(oldDiff(a, b, expanded));
    });
  });
});

test.skip('collapses big diffs to patch format', () => {
  const arg1 = {test: [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]};
  const arg2 = {test: [1, 2, 3, 4, 5, 6, 7, 8, 10, 9]};
  const result = diff(arg1, arg2, unexpanded);
  expect(result).toBe(oldDiff(arg1, arg2, unexpanded));
});

describe.skip('context', () => {
  const testDiffContextLines = (contextLines?: number) => {
    const validContextLines =
      typeof contextLines === 'number' &&
      Number.isSafeInteger(contextLines) &&
      contextLines >= 0;

    test(`number of lines: ${
      typeof contextLines === 'number' ? contextLines : 'undefined'
    } ${validContextLines ? '' : '(5 default)'}`, () => {
      const options = {
        ...optionsCounts,
        contextLines,
        expand: false,
      };
      if (!validContextLines) {
        options.patchColor = chalk.dim;
      }

      const arg1 = {test: [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]};
      const arg2 = {test: [1, 2, 3, 4, 5, 6, 7, 8, 10, 9]};

      const result = diff(arg1, arg2, options);
      expect(result).toEqual(oldDiff(arg1, arg2, options));
    });
  };

  testDiffContextLines(-1); // (5 default)
  testDiffContextLines(0);
  testDiffContextLines(1);
  testDiffContextLines(2);
  testDiffContextLines(3.1); // (5 default)
  testDiffContextLines(); // (5 default)
});
