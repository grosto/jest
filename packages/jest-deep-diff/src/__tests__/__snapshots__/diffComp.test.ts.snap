// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`collapses big diffs to patch format 1`] = `
<g>- Expected</>
<r>+ Received</>

<d>  Object {</>
<d>    "test": Array [</>
<d>      1,</>
<d>      2,</>
<d>      3,</>
<d>      4,</>
<d>      5,</>
<d>      6,</>
<d>      7,</>
<d>      8,</>
<g>-     9,</>
<r>+     10,</>
<g>-     10,</>
<r>+     9,</>
<d>    ],</>
<d>  }</>
`;

exports[`color of text (expanded) 1`] = `
<g>- Expected</>
<r>+ Received</>

<d>  Object {</>
<d>    "searching": "",</>
<g>-   "sorting": Object {</>
<g>      "descending": false,</>
<g>      "fieldKey": "what",</>
<g>    },</>
<r>+   "sorting": Array [</>
<r>      Object {</>
<r>        "descending": false,</>
<r>        "fieldKey": "what",</>
<r>      },</>
<r>    ],</>
<d>  }</>
`;

exports[`context number of lines: -1 (5 default) 1`] = `
<g>- Expected</>
<r>+ Received</>

<d>  Object {</>
<d>    "test": Array [</>
<d>      1,</>
<d>      2,</>
<d>      3,</>
<d>      4,</>
<d>      5,</>
<d>      6,</>
<d>      7,</>
<d>      8,</>
<g>-     9,</>
<r>+     10,</>
<g>-     10,</>
<r>+     9,</>
<d>    ],</>
<d>  }</>
`;

exports[`context number of lines: 0  1`] = `
<g>- Expected</>
<r>+ Received</>

<d>  Object {</>
<d>    "test": Array [</>
<d>      1,</>
<d>      2,</>
<d>      3,</>
<d>      4,</>
<d>      5,</>
<d>      6,</>
<d>      7,</>
<d>      8,</>
<g>-     9,</>
<r>+     10,</>
<g>-     10,</>
<r>+     9,</>
<d>    ],</>
<d>  }</>
`;

exports[`context number of lines: 1  1`] = `
<g>- Expected</>
<r>+ Received</>

<d>  Object {</>
<d>    "test": Array [</>
<d>      1,</>
<d>      2,</>
<d>      3,</>
<d>      4,</>
<d>      5,</>
<d>      6,</>
<d>      7,</>
<d>      8,</>
<g>-     9,</>
<r>+     10,</>
<g>-     10,</>
<r>+     9,</>
<d>    ],</>
<d>  }</>
`;

exports[`context number of lines: 2  1`] = `
<g>- Expected</>
<r>+ Received</>

<d>  Object {</>
<d>    "test": Array [</>
<d>      1,</>
<d>      2,</>
<d>      3,</>
<d>      4,</>
<d>      5,</>
<d>      6,</>
<d>      7,</>
<d>      8,</>
<g>-     9,</>
<r>+     10,</>
<g>-     10,</>
<r>+     9,</>
<d>    ],</>
<d>  }</>
`;

exports[`context number of lines: 3.1 (5 default) 1`] = `
<g>- Expected</>
<r>+ Received</>

<d>  Object {</>
<d>    "test": Array [</>
<d>      1,</>
<d>      2,</>
<d>      3,</>
<d>      4,</>
<d>      5,</>
<d>      6,</>
<d>      7,</>
<d>      8,</>
<g>-     9,</>
<r>+     10,</>
<g>-     10,</>
<r>+     9,</>
<d>    ],</>
<d>  }</>
`;

exports[`context number of lines: undefined (5 default) 1`] = `
<g>- Expected</>
<r>+ Received</>

<d>  Object {</>
<d>    "test": Array [</>
<d>      1,</>
<d>      2,</>
<d>      3,</>
<d>      4,</>
<d>      5,</>
<d>      6,</>
<d>      7,</>
<d>      8,</>
<g>-     9,</>
<r>+     10,</>
<g>-     10,</>
<r>+     9,</>
<d>    ],</>
<d>  }</>
`;
